bookshelf/src/DataModel/Datastore.php

```diff
@@ -0,0 +1,246 @@
+        return $key->getPath()[0]->getId();
+    }
+
+    public function read($id)
+    {
+        $key = $this->createKey($id);
+        $request = new \Google_Service_Datastore_LookupRequest([
+            'keys' => [$key]
+        ]);
+
+        $response = $this->datastore->datasets->
+            lookup($this->datasetId, $request);
+
+        /** @var \Google_Service_Datastore_QueryResultBatch $batch */
+        if ($found = $response->getFound()) {
+            $book = $this->propertiesToBook($found[0]['entity']['properties']);
+            $book['id'] = $id;
+
+            return $book;
+        }
+
+        return false;
+    }
+
+    public function update($book)
+    {
+        $this->verifyBook($book);
+
+        if (!isset($book['id'])) {
+            throw new InvalidArgumentException('Book must have an "id" attribute');
+        }
+
+        $key = $this->createKey($book['id']);
+        $properties = $this->bookToProperties($book);
+
+        $entity = new \Google_Service_Datastore_Entity([
+            'key' => $key,
+            'properties' => $properties
+        ]);
+
+        $request = new \Google_Service_Datastore_CommitRequest([
+            'mode' => 'NON_TRANSACTIONAL',
+            'mutation' => [
+                'update' => [$entity]
+            ]
+        ]);
+
+        $response = $this->datastore->datasets->commit($this->datasetId, $request);
+
+        // return the number of updated rows
+        return 1;
+    }
+
+    public function delete($id)
+    {
+        $key = $this->createKey($id);
+
+        $request = new \Google_Service_Datastore_CommitRequest([
+            'mode' => 'NON_TRANSACTIONAL',
+            'mutation' => [
+                'delete' => [$key]
+            ]
+        ]);
+
+        $response = $this->datastore->datasets->commit($this->datasetId, $request);
+
+        return true;
+    }
+
+    protected function createKey($id = null)
+    {
+        // If we have an ID already, create a "Key" object
+        if ($id) {
+            $key = new \Google_Service_Datastore_Key([
+                'path' => [
+                    [
+                        'id' => $id,
+                        'kind' => 'Book'
+                    ],
+                ]
+            ]);
+
+            return $key;
+        }
+
+        // Otherwise, request a unique ID for a new object
+        $keyRequest = new \Google_Service_Datastore_AllocateIdsRequest([
+            'keys' => [
+                [
+                    'path' => [
+                        [
+                            'kind' => 'Book',
+                        ]
+                    ]
+                ]
+            ]
+        ]);
+
+        $uniqueId = $this->datastore->datasets->allocateIds($this->datasetId, $keyRequest);
+
+        return $uniqueId->getKeys()[0];
+    }
+
+    private function verifyBook($book)
+    {
+        if ($invalid = array_diff_key($book, $this->columns)) {
+            throw new \Exception(sprintf(
```

Comment:
Yes but it *would* be better to be more specific. In this case, `\InvalidArgumentException` would be more precise.